README.md
[![codecov](https://codecov.io/gh/orolia/prisma/branch/develop/graph/badge.svg?token=isXJvglRlX)](https://codecov.io/gh/orolia/prisma)
[![AWS Codebuild](https://codebuild.us-east-1.amazonaws.com/badges?uuid=eyJlbmNyeXB0ZWREYXRhIjoiZTViQWpLb2Zib1R2R2JoVVhzd0VmUW1LekpNUEhYNlNjSjBpRlJPTHVpYjhZcDFLVnZ1UkkvRWFIUXoxc1NJd0s2bHF0UExBMUdaUC85NkhVeTlxeXZrPSIsIml2UGFyYW1ldGVyU3BlYyI6InJKelFVYllMZTJsNll1Mm4iLCJtYXRlcmlhbFNldFNlcmlhbCI6MX0%3D&branch=develop)](https://console.aws.amazon.com/codesuite/codebuild/projects/prisma/history?region=us-east-1)

# Prisma C2 Platform

## Summary

The Prisma repository contains one main projects:

* TMS: Service daemons used for the collection and distribution of sensor data.

Documentation is located in the `docs` directory. Run `make serve-docs` to serve the docs locally at `127.0.0.1:8000`.

You may need to run the following two lines to get the docs building:

```bash tab="macOS"
brew install mkdocs
pip3 install mkdocs-material pymdown-extensions
```

```bash tab="Ubuntu"
sudo apt install mkdocs
pip install mkdocs-material pymdown-extensions
```

## TMS

### Development Setup

The easiest way to get started with TMS development is to use Vagrant to setup
a virtual machine. Install the following:

* [VirtualBox](https://www.virtualbox.org/wiki/VirtualBox)
* [Vagrant](https://www.vagrantup.com/)

Create the virtual machine by executing the following command from the
repository root:

```bash
vagrant up
```
For running one more machine use the next snippet:
```bash
vagrant up cc2
```

### Building

You can now edit the source files on your local system with your favorite
editor and then compile in the virtual machine with:

```bash
vagrant ssh
cd ~/go/src/prisma
make
```
For another machine:
```bash
vagrant ssh cc2
cd ~/go/src/prisma
make
```

Once complete, the entire installation will available under `bin`.

TMS executables should be available in the standard path. Symlinks are created
to the artifacts built in the `bin` directory.

When changing the protobuf files, use `make protobuf` to regenerate the
related source code. The default Makefile target does not perform this step.

### Makefile targets

| Target      | Action                                             |
|-------------|----------------------------------------------------|
| all         | Compiles the server code and documentation
| init        | Prepare the project to be run
| docs        | Generate static documentation site
| serve-docs  | Generate documentation site server on 127.0.0.1
| version     | Show current version of project
| protobuf    | Generate source code from protobuf definitions
| format      | Automatically format the source code
| test        | Run the test suite
| testbed     | Create deb package
| vet         | Check for programming or style errors
| dist        | Create a debian package using the existing build binaries
| clean       | Remove all generated files
| integration | Run the integration test suite
| acceptance  | Is the same as integration
| demo        | Run the demo data load script (tmsd must be running)
| cover       | Run the test suit and coverage outputs .cover.html file
| lint        | Run linter with errcheck, unused, deadcode, and gosec enabled. output stdout.


#### macOS Development Setup
Running on native macOS

Pros
*  Faster compile
*  Debugger

Cons
*  Complex setup
*  Greater configuration variance

Compile
```
brew install ejdb
```

**MongoDB**

Native
```
brew install mongodb
mkdir -p ~/data/db
mongod --dbpath ~/data/db --port 8201 --replSet replocal
mongo localhost:8201 etc/mongodb/replication.js
mongo localhost:8201/trident etc/mongodb/schema/trident.js
mongo localhost:8201/aaa etc/mongodb/schema/aaa.js
```

Docker (not working, replicaset port)
```
docker run --name prisma-mongo -d -p 127.0.0.1:8201:8201 mongo --replSet replocal
mongo localhost:8201 etc/mongodb/replication.js
mongo localhost:8201/trident etc/mongodb/schema/trident.js
mongo localhost:8201/aaa etc/mongodb/schema/aaa.js
```

Consul
```
brew install consul
consul agent -dev -bind 127.0.0.1 -client 127.0.0.1
```

Redis
```
brew install redis
redis-server /usr/local/etc/redis.conf
```

## Running

Use [tmsd](tms/cmd/daemons/tmsd/README.md) to manage the TMS processes:

| Command            | Use                              |
|--------------------|----------------------------------|
| `tmsd --start &`   | Start all processes              |
| `tmsd --info`      | Verify the processes are running |
| `tmsd --stop`      | Stop all processes               |
| `tmsd --help`      | For additional information       |

### REST Documentation

#### Postman (Preferred)

We currently provide two types of REST API documentation. For the most up to date
and best documentation experience, run POSTMAN app or see the POSTMAN hosted documentation at [Orolia POSTMAN C2 Documentation Site.](https://orolia-prisma-c2.postman.co/collections/352230-e1e86aaf-b2c3-4e3b-b718-a74f4b7a0281?workspace=44d15d8d-0a82-4c95-8e74-d196d18dd868)

#### Swagger

Additionally we also provide swagger documentation.

To view the JSON for the RESTful API:
http://localhost:8080/api/v2/apidocs.json

To view with the swagger UI:

1. [Install latest docker]( https://store.docker.com/editions/community/docker-ce-desktop-mac)
2. run
  ```docker pull swaggerapi/swagger-ui;
  docker run -p 80:8080 swaggerapi/swagger-ui
  ```
3. Goto - http://localhost/?url=https://localhost:8080/api/v2/apidocs.json

### Configuration

Processes managed by tmsd are configured in `/etc/trident/tmsd.conf` or
`/etc/trident/tmsd_cc2.conf` for remote machine.

On a clean install, no configuration is currently provided, but you can use the vagrant config as the basis for creating new configuration files.

### Logging

TMS processes use the system logger and are directed to `/var/log/tmsd.log`.
When running a process manually, use the `--stdlog` option to redirect
log messages to the console.

The minimum default log level is "info" for all the processes. Change the
logging level using `--log LEVEL` flag. The log levels are the same as those
used by syslog.

Tracer objects log at the alert level and are only enabled when passing in
`--trace NAME` on the command line where NAME is the given name of the tracer.

## Client

See https://github.com/orolia/prisma-ui for client project.

### Documentation for backend
To see documentation about packages you can run godoc
```bash
gorun -http :6060
```
Go to [http://localhost:6060/pkg/prisma]

### Documentations for daemons
- **[tanalyzed](tms/cmd/daemons/tanalyzed/README.md)**
- **[tauthd](tms/cmd/daemons/tauthd/README.md)**
- **[tdatabased](tms/cmd/daemons/tdatabased/README.md)**
- **[tfleetd](tms/cmd/daemons/tfleetd/README.md)**
- **[tgwad](tms/cmd/daemons/tgwad/README.md)**
- **[tmccd](tms/cmd/daemons//tmccd/README.md)**
- **[tmsd](tms/cmd/daemons/tmsd/README.md)**
- **[tnafexportd](tms/cmd/daemons/tnafexportd/README.md)**
- **[tnoid](tms/cmd/daemons/tnoid/README.md)**
- **[torbcommd](tms/cmd/daemons/torbcommd/README.md)**
- **[treportd](tms/cmd/daemons/treportd/README.md)**
- **[treportd](tms/cmd/daemons/treportd/README.md)**
- **[tselfsign](tms/cmd/daemons/tselfsign/README.md)**
- **[tspiderd](tms/cmd/daemons/tspiderd/README.md)**
- **[twebd](tms/cmd/daemons/twebd/README.md)**
- **[tauthd](tms/cmd/daemons/tauthd/README.md)**
- **[twebd](tms/cmd/daemons/twebd/README.md)**

### Documentations for tools
- **[datagramdownlinkrequest](tms/cmd/tools/datagramdownlinkrequest/README.md)**
- **[generator](tms/cmd/tools/generator/README.md)**
- **[nmea-lib-gen](tms/cmd/tools/nmea-lib-gen/README.md)**
- **[omnicom-iridium](tms/cmd/tools/omnicom-iridium/README.md)**
- **[omnicom-lib-gen](tms/cmd/tools/omnicom-lib-gen/README.md)**
- **[rsm](tms/cmd/tools/rsm/README.md)**
- **[sarsat-start](tms/cmd/tools/sarsat-start/README.md)**
- **[tdemo](tms/cmd/tools/tdemo/README.md)**
- **[tloggernmea](tms/cmd/tools/tloggernmea/README.md)**
- **[tmccrd](tms/cmd/tools/tmccrd/README.md)**
- **[tmongo](tms/cmd/tools/tmongo/README.md)**
- **[tping](tms/cmd/tools/tping/README.md)**
- **[tportald](tms/cmd/tools/tportald/README.md)**
- **[tsimulator](tms/cmd/tools/tsimulator/README.md)**
- **[tstarter-tnoids](tms/cmd/tools/tstarter-tnoids/README.md)**
- **[twatch](tms/cmd/tools/twatch/README.md)**

## Other topics

* [Enabling SSL/TLS security](https://documentation.mcmurdo.io/installation/ssl/)


CHANGELOG.md
# PRISMA Changelog

For client changes see [https://github.com/orolia/prisma-ui/blob/develop/CHANGELOG.md](https://github.com/orolia/prisma-ui/blob/develop/CHANGELOG.md)

More release information and downloads are at [https://github.com/orolia/prisma/releases](https://github.com/orolia/prisma/releases)

## 1.8.0 release 2 (Oct 19, 2020)

This second version of 1.8.0 fixes a critical ADS-B bug.

### TMS Changes

[FIX] Create track and registry ID using callsign or Code S. [PR#110](https://github.com/mcmsar/prisma/pull/110)

### TMS Changes

* [NEW] Add support for TEST USER PROTOCOL decoding. [PR#98](https://github.com/mcmsar/prisma/pull/98)
* [NEW] Add vts support. [PR#100](https://github.com/mcmsar/prisma/pull/100)
* [NEW] Add support for beacon ID with orbitography Protocol data. [PR#103](https://github.com/mcmsar/prisma/pull/103)
* [NEW] SIT 915 support. [PR#99](https://github.com/mcmsar/prisma/pull/99)
* [NEW] Added support for the AirNav RadarBox data ingestion. [PR#97](https://github.com/mcmsar/prisma/pull/97)
* [NEW] Add RLS support. [PR#95](https://github.com/mcmsar/prisma/pull/95)

## 1.8.0 release 1 (Oct 19, 2020)

This is primarily a release for delivering RCC system for Qatar to replace legacy RCC.

### TMS Changes

* [NEW] Add support for TEST USER PROTOCOL decoding. [PR#98](https://github.com/mcmsar/prisma/pull/98)
* [NEW] Add vts support. [PR#100](https://github.com/mcmsar/prisma/pull/100)
* [NEW] Add support for beacon ID with orbitography Protocol data. [PR#103](https://github.com/mcmsar/prisma/pull/103)
* [NEW] SIT 915 support. [PR#99](https://github.com/mcmsar/prisma/pull/99)
* [NEW] Added support for the AirNav RadarBox data ingestion. [PR#97](https://github.com/mcmsar/prisma/pull/97)
* [NEW] Add RLS support. [PR#95](https://github.com/mcmsar/prisma/pull/95)

## 1.8 release candidate 1 (Oct 1, 2020)

This is primarily a release for delivering RCC system for Qatar to replace legacy RCC.

### TMS Changes

* [NEW] Add support for TEST USER PROTOCOL decoding. [PR#98](https://github.com/mcmsar/prisma/pull/98)
* [NEW] Add vts support. [PR#100](https://github.com/mcmsar/prisma/pull/100)
* [NEW] Add support for beacon ID with orbitography Protocol data. [PR#103](https://github.com/mcmsar/prisma/pull/103)
* [NEW] SIT 915 support. [PR#99](https://github.com/mcmsar/prisma/pull/99)
* [NEW] Added support for the AirNav RadarBox data ingestion. [PR#97](https://github.com/mcmsar/prisma/pull/97)
* [NEW] Add RLS support. [PR#95](https://github.com/mcmsar/prisma/pull/95)

### Developer Changes.

* [NEW] Add configuration for ubuntu bionic64 to the Vagrantfile. [PR#101](https://github.com/mcmsar/prisma/pull/101)
* [CHG] Changing proto-gen-go version from latest to v1.3.0 and protoc version to v3.11.4. [PR#96](https://github.com/mcmsar/prisma/pull/96)


## 1.7.7 (Jan 29, 2020)

This is primarily a release for delivering couple new features to Singapore RCC.

### TMS Changes.

* [NEW] Add parameter to allow deleted log entries for incidents. [PR#83](https://github.com/mcmsar/prisma/pull/83)
* [NEW] Add policy to prohibit user id as password. [PR#84](https://github.com/mcmsar/prisma/pull/84)

### Developer Changes.

* [NEW] add parameter "first" to track request to get the first target instead of the last. [PR#82](https://github.com/mcmsar/prisma/pull/82)
* [FIX] Corrected longitude validation in twebd. [PR#87](https://github.com/mcmsar/prisma/pull/87)
* [FIX] Add new endpoint, history-database, that has the behavior of the hold history endpoint used for tooltip and info panels. [PR#88](https://github.com/mcmsar/prisma/pull/88)
* [CHG]: Update TrackID hash function for SARSAT beacons to use only HexID as a seed. [PR#89](https://github.com/mcmsar/prisma/pull/89)

## 1.7.6 (May 24, 2019)

This is primarily a technical debt release which includes many dependency updates, documentation changes, and build system updates. This release is not intended to be released to customers. 

### TMS Changes

* [NEW] When a track is added to an incident, it will not longer be removed from the map until the incident is closed. [PR #48](https://github.com/orolia/prisma/pull/48)
* [NEW] Can now set time to live indexes on data stored in mongo so data can automatically be purged after an amount of time. [PR #38](https://github.com/orolia/prisma/pull/38)
* [CHG] Revert back to `token` property from `password` for tauthd session login. [PR #36](https://github.com/orolia/prisma/pull/36)
* [CHG] Sites can now be retrieved by mongo `_id` and site id. [PR #29](https://github.com/orolia/prisma/pull/29), [PR #41](https://github.com/orolia/prisma/pull/41)
* [CHG] Disabled insecure ciphers for TLS configurations [PR #42](https://github.com/orolia/prisma/pull/42)
* [FIX] When certificates are rolled over, daemons no longer need to be restarted to pick up the new certificates. [PR #40](https://github.com/orolia/prisma/pull/40)
* [FIX] Fixed critical issue where systems not using x509 certificates to authenticate mongo wouldn't start. [PR #35](https://github.com/orolia/prisma/pull/35)
* [FIX] Fixed issue with NAF parsing failing due to empty voltage on systems deployed for the Ministry of Fisheries. [PR #10](https://github.com/orolia/prisma/pull/10)
* [FIX] Fixed some links in the documentation site and added new documentation for docker, replication, and other updates. [PR #45](https://github.com/orolia/prisma/pull/45), [PR #46](https://github.com/orolia/prisma/pull/46), [PR #47](https://github.com/orolia/prisma/pull/47)
* [FIX] Fixed an issue where trackex wasn't loading correctly and causing tracks to disappear before they should. [PR #43](https://github.com/orolia/prisma/pull/43)

### Developer Changes

* [NEW] Finished moving the repo to github. Documentation now links to github repos. [PR #2](https://github.com/orolia/prisma/pull/2), [PR #17](https://github.com/orolia/prisma/pull/17) 
* [NEW] Added golinter [PR #31](https://github.com/orolia/prisma/pull/31)

tms/sar/bits.go
//Package sar implements specifications for the sarsat beacon identifier.
//The specifications can be found here:
// https://team.technosci.com:8442/browse/CONV-801
package sar

import (
	fmt "fmt"
	"strings"
)

// Reuse the structure for readers and writers
type bits struct {
	bytes []byte
	pos   int // index into the byte array
	nbits int // bits consumed in the current byte
}

func (b *bits) String() string {
	fbytes := make([]string, len(b.bytes))
	for i, it := range b.bytes {
		fbytes[i] = fmt.Sprintf("%08b", it)
	}
	return strings.Join(fbytes, " ")
}

// BitReader reads bits from a byte array. Reads start at the first byte
// in the array at the MSB and works towards the LSB.
type BitReader struct {
	bits
}

// NewBitReader creates a new reader for the given series of bytes. The value
// passed in will be mutated, make a copy if necessary.
func NewBitReader(data []byte) *BitReader {
	b := &BitReader{
		bits: bits{
			bytes: data,
		},
	}
	return b
}

func (b *BitReader) Read() uint64 {
	v := b.bytes[b.pos] & 0x80
	b.bytes[b.pos] <<= 1
	b.nbits++
	if b.nbits == 8 {
		b.nbits = 0
		b.pos++
	}
	if v != 0 {
		return 1
	}
	return 0
}

func (b *BitReader) ReadN(n int) uint64 {
	var v uint64
	for i := 0; i < n; i++ {
		v <<= 1
		v |= b.Read()
	}
	return v
}

func (b *BitReader) SkipN(n int) {
	_ = b.ReadN(n)
}

func (b *BitReader) ReadBaudot() (rune, error) {
	return b.readBaudot(baudotSixBit)
}

func (b *BitReader) ReadBaudot5() (rune, error) {
	return b.readBaudot(baudotFiveBit)
}

func (b *BitReader) readBaudot(size baudotSize) (rune, error) {
	var code int
	switch size {
	case baudotSixBit:
		code = int(b.ReadN(6))
	case baudotFiveBit:
		code = int(b.ReadN(5)) | 32 // always set 6th bit 0b100000
	default:
		return 0, fmt.Errorf("invalid size: %v", size)
	}
	ch, ok := baudotToRune[code]
	if !ok {
		return 0, fmt.Errorf("invalid baudot code: %v", code)
	}
	return ch, nil
}

func (b *BitReader) ReadBaudotN(n int) (string, error) {
	return b.readBaudotN(n, baudotSixBit)
}

func (b *BitReader) ReadBaudot5N(n int) (string, error) {
	return b.readBaudotN(n, baudotFiveBit)
}

func (b *BitReader) readBaudotN(n int, size baudotSize) (string, error) {
	runes := make([]rune, n)
	for i := 0; i < n; i++ {
		ch, err := b.readBaudot(size)
		if err != nil {
			return "", err
		}
		runes[i] = ch
	}
	return string(runes), nil
}

// BitWriter writes bits to a byte array. Writes start at the first byte
// in the array at the MSB and works towards the LSB.

type BitWriter struct {
	bits
}

func (b *BitWriter) Write(val int) *BitWriter {
	if b.bytes == nil {
		b.bytes = make([]byte, 1)
		b.nbits = 8
	} else if b.nbits == 0 {
		b.bytes = append(b.bytes, byte(0))
		b.nbits = 8
		b.pos++
	}

	b.nbits--
	if val != 0 {
		val = 1 << uint(b.nbits)
	}
	b.bytes[b.pos] |= byte(val)
	return b
}

func (b *BitWriter) WriteN(n int, val int) *BitWriter {
	for i := n - 1; i >= 0; i-- {
		mask := 1 << uint(i)
		b.Write(val & mask)
	}
	return b
}

func (b *BitWriter) WriteZeros(n int) *BitWriter {
	for i := 0; i < n; i++ {
		b.Write(0)
	}
	return b
}

func (b *BitWriter) WriteBaudot(r rune) error {
	return b.writeBaudot(r, baudotSixBit)
}

func (b *BitWriter) WriteBaudot5(r rune) error {
	return b.writeBaudot(r, baudotFiveBit)
}

func (b *BitWriter) writeBaudot(r rune, size baudotSize) error {
	code, ok := runeToBaudot[r]
	if !ok {
		return fmt.Errorf("no encoding for rune: %v", r)
	}
	switch size {
	case baudotSixBit:
		b.WriteN(6, code)
	case baudotFiveBit:
		b.WriteN(5, code)
	default:
		return fmt.Errorf("invalid size: %v", size)
	}
	return nil
}

func (b *BitWriter) WriteBaudotN(n int, str string) error {
	return b.writeBaudotN(n, str, baudotSixBit)
}

func (b *BitWriter) WriteBaudot5N(n int, str string) error {
	return b.writeBaudotN(n, str, baudotFiveBit)
}

func (b *BitWriter) writeBaudotN(n int, str string, size baudotSize) error {
	pad := n - len(str)
	for i := 0; i < pad; i++ {
		if err := b.writeBaudot(' ', size); err != nil {
			return err
		}
	}
	for _, ch := range str {
		if err := b.writeBaudot(rune(ch), size); err != nil {
			return err
		}
	}
	return nil
}

// Modified-Baudot is normally six bits but the serial standard location
// protocol message with a aircraft operator desginator uses a five bit
// variant
type baudotSize bool

const (
	baudotSixBit  baudotSize = true
	baudotFiveBit            = false
)

var baudotToRune = map[int]rune{
	56: 'A', // 0b111000
	51: 'B', // 0b110011
	46: 'C', // 0b101110
	50: 'D', // 0b110010
	48: 'E', // 0b110000
	54: 'F', // 0b110110
	43: 'G', // 0b101011
	37: 'H', // 0b100101
	44: 'I', // 0b101100
	58: 'J', // 0b111010
	62: 'K', // 0b111110
	41: 'L', // 0b101001
	39: 'M', // 0b100111
	38: 'N', // 0b100110
	35: 'O', // 0b100011
	45: 'P', // 0b101101
	61: 'Q', // 0b111101
	42: 'R', // 0b101010
	52: 'S', // 0b110100
	33: 'T', // 0b100001
	60: 'U', // 0b111100
	47: 'V', // 0b101111
	57: 'W', // 0b111001
	55: 'X', // 0b110111
	53: 'Y', // 0b110101
	49: 'Z', // 0b110001
	36: ' ', // 0b100100
	24: '-', // 0b011000
	23: '/', // 0b010111
	13: '0', // 0b001101
	29: '1', // 0b011101
	25: '2', // 0b011001
	16: '3', // 0b010000
	10: '4', // 0b001010
	1:  '5', // 0b000001
	21: '6', // 0b010101
	28: '7', // 0b011100
	12: '8', // 0b001100
	3:  '9', // 0b000011
}

var runeToBaudot = map[rune]int{}

func init() {
	for code, r := range baudotToRune {
		runeToBaudot[r] = code
	}
}
